sklearn/utils/_set_output.py

```diff
@@ -1,68 +1,187 @@
+
+        Returns
+        -------
+        is_supported_container : bool
+            True if X is a supported container.
+        """
+
+    def rename_columns(self, X, columns):
+        """Rename columns in `X`.
+
+        Parameters
+        ----------
+        X : container
+            Container which columns is updated.
+
+        columns : ndarray of str
+            Columns to update the `X`'s columns with.
+
+        Returns
+        -------
+        updated_container : container
+            Container with new names.
+        """
+
+    def hstack(self, Xs):
+        """Stack containers horizontally (column-wise).
+
+        Parameters
+        ----------
+        Xs : list of containers
+            List of containers to stack.
+
+        Returns
+        -------
+        stacked_Xs : container
+            Stacked containers.
+        """
+
+
+class PandasAdapter:
+    container_lib = "pandas"
+
+    def create_container(self, X_output, X_original, columns):
+        pd = check_library_installed("pandas")
+        columns = get_columns(columns)
+        index = X_original.index if isinstance(X_original, pd.DataFrame) else None
+
+        if isinstance(X_output, pd.DataFrame):
+            if columns is not None:
+                X_output.columns = columns
+            return X_output
+
+        return pd.DataFrame(X_output, index=index, columns=columns, copy=False)
+
+    def is_supported_container(self, X):
+        pd = check_library_installed("pandas")
+        return isinstance(X, pd.DataFrame)
+
+    def rename_columns(self, X, columns):
+        return X.rename(columns=dict(zip(X.columns, columns)))
+
+    def hstack(self, Xs):
+        pd = check_library_installed("pandas")
+        return pd.concat(Xs, axis=1)
+
+
+class PolarsAdapter:
+    container_lib = "polars"
+
+    def create_container(self, X_output, X_original, columns):
+        pl = check_library_installed("polars")
+        columns = get_columns(columns)
 
-    return pd.DataFrame(data_to_wrap, index=index, columns=columns, copy=False)
+        if isinstance(X_output, pl.DataFrame):
+            if columns is not None:
+                return self.rename_columns(X_output, columns)
+            return X_output
+
+        if isinstance(columns, np.ndarray):
+            columns = columns.tolist()
+
+        return pl.DataFrame(X_output, schema=columns)
+
+    def is_supported_container(self, X):
+        pl = check_library_installed("polars")
+        return isinstance(X, pl.DataFrame)
+
+    def rename_columns(self, X, columns):
+        return X.rename(dict(zip(X.columns, columns)))
+
+    def hstack(self, Xs):
+        pl = check_library_installed("polars")
+        return pl.concat(Xs, how="horizontal")
+
+
+class ContainerAdaptersManager:
+    def __init__(self):
+        self.adapters = {}
+
+    @property
+    def supported_outputs(self):
+        return {"default"} | set(self.adapters)
+
+    def register(self, adapter):
```

Comment:
```suggestion
    def register(self, adapter):
        """Register a container adapter.

        Parameters
        ----------
        adapter : :class:`ContainerAdapterProtocol`
            A container adapter that follows the protocol defined by
            :class:`ContainerAdapterProtocol`.
```

In this regards, do we want to enforce something like:

```python
if not isintance(adapter, ContainerAdapterProtocol):
    raise TypeError(
        "The adapter does not follow the ContainerAdapterProtocol."
    )
```

Or we want to be lenient?